### PR TITLE
Add proxy support to webkitscmpy's GitHub client

### PR DESCRIPTION
**Pull request by alancoon**
#### 74c767d376d0f9aaf014ea36de0cb3d043e97cd1
<pre>
Add proxy support to webkitscmpy&apos;s GitHub client
<a href="https://bugs.webkit.org/show_bug.cgi?id=241031">https://bugs.webkit.org/show_bug.cgi?id=241031</a>
rdar://93129765

Reviewed by Jonathan Bedard.

Allow for the passing of proxy settings to the GitHub client. Use
requests.Session to pass the proxy settings to all requests made.

Additionally pass this session to the Tracker data member, for
github.com networking requests made at the Tracker level.

Version bumps webkitbugspy and webkitscmpy.

* Tools/Scripts/libraries/webkitbugspy/setup.py: version bump to 0.6.3
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: version bump to 0.6.3
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
Add a requests.Session member to the class that can be passed into the constructor.
Call all networking requests through the session.
(Tracker.__init__):
(Tracker):
(Tracker.credentials.validater):
* Tools/Scripts/libraries/webkitscmpy/setup.py: version bump to 4.15.4
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: version bump to 4.15.4
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
Add a proxies argument to the constructor. Will pass proxies to a request.Session member
that is used to perform networking requests to github.com from this class.
This session is passed to the Tracker member in its constructor (see webkitbugspy.github changes).
(GitHub.PRGenerator.create):
(GitHub.PRGenerator.update):
(GitHub.__init__):
(GitHub.request):
(GitHub._count_for_ref):
(GitHub._branches_for):

Canonical link: <a href="https://commits.webkit.org/251349@main">https://commits.webkit.org/251349@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295325">https://svn.webkit.org/repository/webkit/trunk@295325</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Review by JonWBedard — CHANGES_REQUESTED** [review]
Need to rebase this on [251070@main](https://commits.webkit.org/251070@main)

**Comment by alancoon** [comment]
> Need to rebase this on [251070@main](https://commits.webkit.org/251070@main)

Those changes were to `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py`, whereas these changes alter `Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py`.

**Comment by JonWBedard** [comment]
> > Need to rebase this on [251070@main](https://commits.webkit.org/251070@main)
> 
> Those changes were to `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py`, whereas these changes alter `Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py`.

I think the `webkitbugspy/github.py` stuff needs the proxy too, since `webkitbugspy/github.py` is the base class for `webkitscmpy/remote/git_hub.py`

**Comment by alancoon** [comment]
> > > Need to rebase this on [251070@main](https://commits.webkit.org/251070@main)
> > 
> > 
> > Those changes were to `Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py`, whereas these changes alter `Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py`.
> 
> I think the `webkitbugspy/github.py` stuff needs the proxy too, since `webkitbugspy/github.py` is the base class for `webkitscmpy/remote/git_hub.py`

I see, depending on how the client uses the `github.Tracker` class, they may wind up calling requests to GitHub.com, in which case we'd need to also proxy the calls

**Comment by JonWBedard** [comment]
> > ...
> 
> I see, depending on how the client uses the `github.Tracker` class, they may wind up calling requests to GitHub.com, in which case we'd need to also proxy the calls

Right, the real concern I have is the fact that PRs are actually implemented as issues under the hood, so there are a lot of scripts that interact with GitHub Issues when you wouldn't expect them to.

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py [hunk 1]
```diff
@@ -65,10 +65,11 @@ def default(context, obj):
                 raise TypeError('Cannot invoke parent class when classmethod')
             return super(Tracker.Encoder, context).default(obj)
 
-
-    def __init__(self, url, users=None, res=None, component_color=DEFAULT_COMPONENT_COLOR, version_color=DEFAULT_VERSION_COLOR):
+    def __init__(self, url, users=None, res=None, component_color=DEFAULT_COMPONENT_COLOR,
+                 version_color=DEFAULT_VERSION_COLOR, session=None):
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py]: If we're going to add a newline here, can we do something like this:
```
def __init__(
    self, url, users=None, res=None,
    component_color=DEFAULT_COMPONENT_COLOR,
    version_color=DEFAULT_VERSION_COLOR,
    session=None,
):
```

**Review by JonWBedard — APPROVED** [review]
This is a large enough change that we probably need to increment minor versions in:
```
Tools/Scripts/libraries/webkitbugspy/setup.py
Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py
Tools/Scripts/libraries/webkitscmpy/setup.py
Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py
```

**Pull request merged**